plugin/metadata/request.go

```diff
@@ -0,0 +1,212 @@
+	if err != nil {
+		return nil, fmt.Errorf("could not parse EDNS0 data size: %s", err)
+	}
+	start, err := strconv.ParseUint(startStr, 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse EDNS0 start index: %s", err)
+	}
+	end, err := strconv.ParseUint(endStr, 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse EDNS0 end index: %s", err)
+	}
+	if end <= start && end != 0 {
+		return nil, fmt.Errorf("end index should be > start index (actual %d <= %d)", end, start)
+	}
+	if end > size && size != 0 {
+		return nil, fmt.Errorf("end index should be <= size (actual %d > %d)", end, size)
+	}
+	ednsType, ok := stringToEDNS0MapType[dataType]
+	if !ok {
+		return nil, fmt.Errorf("invalid dataType for EDNS0 map: %s", dataType)
+	}
+	ecode := uint16(c)
+	return &edns0Map{name, ecode, ednsType, uint(size), uint(start), uint(end)}, nil
+}
+
+func (m *Metadata) addEDNS0Map(code, name, dataType, sizeStr, startStr, endStr string) error {
+	meta, err := newEDNS0Map(code, name, dataType, sizeStr, startStr, endStr)
+	if err != nil {
+		return err
+	}
+	m.options[meta.code] = append(m.options[meta.code], meta)
+	return nil
+}
+
+// Metadata adds additional standard values to the context.
+func (m *Metadata) Metadata(ctx context.Context, state request.Request) context.Context {
+	return m.fillMetadata(ctx, state)
+}
+
+func (m *Metadata) fillMetadata(ctx context.Context, state request.Request) context.Context {
+
+	m.declareMetadata(ctx, queryName, state.QName())
+	m.declareMetadata(ctx, queryType, dns.Type(state.QType()).String())
+	m.declareMetadata(ctx, clientIP, state.IP())
+	m.declareMetadata(ctx, serverIP, state.LocalIP())
+	m.declareMetadata(ctx, clientPort, state.Port())
+	m.declareMetadata(ctx, serverPort, state.LocalPort())
+	m.declareMetadata(ctx, protocol, state.Proto())
+
+	m.getAttrsFromEDNS0(ctx, state.Req)
+
+	SetValueFunc(ctx, "metadata/"+responseIP, func() string {
+		ip := getRespIP(state.Req)
+		if ip != nil {
+			return ip.String()
+		}
+		return ""
+	})
+	return ctx
+
+}
+
+func (m *Metadata) declareMetadata(ctx context.Context, name string, value string) {
+	SetValueFunc(ctx, "metadata/"+name, func() string { return value })
+}
+
+func (m *Metadata) getAttrsFromEDNS0(ctx context.Context, r *dns.Msg) {
+	o := r.IsEdns0()
+	if o == nil {
+		return
+	}
+
+	for _, opt := range o.Option {
+		optLocal, local := opt.(*dns.EDNS0_LOCAL)
+		if !local {
+			continue
+		}
+		opts, ok := m.options[optLocal.Code]
+		if !ok {
+			continue
+		}
+		m.parseOptionGroup(ctx, optLocal.Data, opts)
+	}
+}
+
+func (m *Metadata) parseOptionGroup(ctx context.Context, data []byte, options []*edns0Map) {
+	for _, option := range options {
+		var value string
+		switch option.dataType {
+		case typeEDNS0Bytes:
+			value = string(data)
+		case typeEDNS0Hex:
+			value = parseHex(data, option)
+		case typeEDNS0IP:
+			ip := net.IP(data)
+			value = ip.String()
+		}
+		if value != "" {
+			m.declareMetadata(ctx, option.name, value)
```

Comment:
Here, for EDNS0 that is much more heavy to handle, we should provide a real lazy initialization.
I mean : do not extract data until it is required by a usage of that data.

And by, the way we should to the same for any field of the Query : 
replace the "extract and cache the value" by "provide the function that extract on demand with a cache"